### PR TITLE
[entropy_src/rtl] Consistent core enable masking

### PR DESCRIPTION
**Pull request by martin-lueker**
The entropy_src core module has many activities that persist after disablement, notably the closing of health_test statistics data and the processing of SHA conditioning outputs.  Prior to this commit there have been many ad-hoc signals to allow these individual processes to finish before clearing the relavant fifos or state machines.  This lack of consistency makes it hard to track how these delays interact.

Ideally the external module interfaces (CSR locks and RNG inputs) would also be closed during these shutdown period to prevent the admission of new RNG data (bad for verification) or new CSR commands (which can cause spurious events due to poorly timed changes in configuration).

This commit creates a new delayed copy of the module enable signals.  The falling edge of this delayed copy signals when it is safe to clear the last few internal FIFOs, and the rising edge signals when the FSM has actually received a pending idle request.

This new enable signal is generated by a new module in order to cleanly encapulate and document the logic and motivation for the delay calculations.

Github Note: This PR requires #16095 for proper DV.

Signed-off-by: Martin Lueker-Boden <martin.lueker-boden@wdc.com>

### CODE REVIEW

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (andreaskurth) [inline hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:86]: Why is this timer implemented as a FIFO instead of a counter?

**Review by andreaskurth — COMMENTED** [review]
Can't comment on the design, but I did not notice plain mistakes in the code

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (mwbranstad) [inline hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:23]: something odd about this sentence.

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (mwbranstad) [inline hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:54]: how was this value chosen? general observation or some specific calculation?

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (mwbranstad) [inline hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:101]: overall impression of this new module is to contain all of the quirky timing controls for the enable/disable sequences, rather than distributed them randomly within the core file. Assuming this is the case, nice!

**Review by mwbranstad — APPROVED** [review]
LGTM, with comments/questions.

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:86]: Just because it is very small, and I (perhaps out of laziness) didn't want to deal with underflow logic.  Would a counter be preferred?

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (martin-lueker) [inline hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:23]:
```suggestion
// This block creates a modified version of the enable pulse which:
```

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:54]: One clock per fifo stage

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:101]: That is the idea, it was getting very unmanageable to document otherwise.

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:54]: Comment added

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (martin-lueker) [inline hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:85]:
```suggestion
  // In flight data monitoring.
  // The `fifo_timer` is a small shift register to count out the maximum number of cycles to wait for the FIFOs to drain.  Since this timer is very small (3 cycle), it is implemented as a shift register to keep the logic simple, though it could easily be  implemented as a counter as well.
```

File: hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv
Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:88]: I've added a comment, letting others know that if the want to replace this with a counter it's fine.

Comment (andreaskurth) [reply hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:88]: I see, makes sense. Thanks for the explanation!

Note that the comment will have to be broken at 100 chars; otherwise lint will fail

Comment (martin-lueker) [reply hw/ip/entropy_src/rtl/entropy_src_enable_delay.sv:88]: Yeah, I figured I'd catch the wrapping in a text editor on my side.

**Comment by tjaychen** [comment]
lgtm, but i'll defer to @mwbranstad

**Comment by vogelpi** [comment]
Update, I just did a rebase on current master as FPGA CI was failing before and a fix for this got merged after @martin-lueker 's last rebase.

**Comment by vogelpi** [comment]
CI now passed. Only the Kokoro check is failing but it's unrelated to this PR. Some config files cannot be found for e.g. for GPIO block-level lint. This is good to go in.

**Pull request merged**
